crates/telio-relay/src/derp/http.rs

```
@@ -89,21 +94,67 @@ pub async fn connect_http_and_start(
+        let stream = timeout(derp_config.timeout, socket.connect(ip)).await??;
+        let addr = PairAddr {
+            local: stream.local_addr()?,
+            remote: stream.peer_addr()?,
+        };
 
-    let stream = timeout(derp_config.timeout, socket.connect(ip)).await??;
+        let connection = match u.scheme() {
+            "http" => {
+                Box::pin(connect_and_start(
+                    stream,
+                    addr,
+                    derp_config.secret_key.clone(),
+                    derp_config.server_keepalives.clone(),
+                    &hostport,
+                    DerpVersion::V2,
+                ))
+                .await
+            }
+            _ => {
+                let config = if derp_config.use_built_in_root_certificates {
+                    let root_store: RootCertStore = TLS_SERVER_ROOTS.iter().cloned().collect();
+
+                    let config = ClientConfig::builder()
+                        .with_root_certificates(root_store)
+                        .with_no_client_auth();
+
+                    TlsConnector::from(Arc::new(config))
+                } else {
+                    TlsConnector::from(Arc::new(rustls_platform_verifier::tls_config()))
+                };
+
+                let server_name =
+                    ServerName::try_from(hostname.clone()).map_err(|_| Error::InvalidServerName)?;
+
+                connect_and_start(
+                    config.connect(server_name, stream).await?,
+                    addr,
+                    derp_config.secret_key.clone(),
+                    derp_config.server_keepalives.clone(),
+                    &hostport,
+                    DerpVersion::V2,
+                )
+                .await
+            }
+        };
+
+        if connection.is_ok() {
+            // We managed to successfully connect to derp v2
+            telio_log_debug!("Successfully connected to derp v2");
+            return connection;
+        }
```

please add a warning if the connection is no ok - completely silently falling back to v1 is probably not what we want.